### PR TITLE
dashboard: add access control interface

### PR DESCRIPTION
**Pull request by dominic**
Adds the ability to grant policy permissions to access tokens and x509 certs 
based on cert subject field. Replaces `Client Token` and `Network Token`
sections in the secondary navigation.

### CODE REVIEW

**Comment by dominic** [comment]
screen shot 2017-04-14 at 5 50 48 pm

screen shot 2017-04-14 at 5 50 49 pm

screen shot 2017-04-14 at 5 51 05 pm

**Comment by dominic** [comment]
Outstanding tasks:

- [ ] Coalesce grants with the same `guard_data` into a single line item
- [ ] Allow coalesced tokens or certs to be "edited" with multiple policies, instead of adding/revoking grants individually
- [ ] Delete access token when all grants removed
- [ ] Nav icon
- [ ] Design polish

File: dashboard/src/features/accessControl/components/AccessControlList.jsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+          </button>
+
+          {tokenList}
+        </div>}
+
+        {this.props.certificatesSelected && <div>
+          <button
+            className={`btn btn-primary ${styles.newBtn}`}
+            onClick={this.props.showAddCertificate}>
+              + Add certificate
```
Comment (jeffomatic) [inline dashboard/src/features/accessControl/components/AccessControlList.jsx]: `Register certificate`?

File: dashboard/src/features/accessControl/components/NewCertificate.jsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+          </button>
+        </FormSection>
+        <FormSection title='Policy'>
+          <SelectField options={policyOptions}
+            title='Policy'
+            hint='Available policies are:
+* `client-readwrite`: full access to the Client API
+* `client-readonly`: access to read-only Client endpoints
+* `network`: access to the Network API
+* `monitoring`: access to monitoring-specific endpoints'
```
Comment (jeffomatic) [inline dashboard/src/features/accessControl/components/NewCertificate.jsx]: Is `internal` so internal that we're not exposing it to the user? I'm not up to speed about how those tokens will be generated and used.

File: dashboard/src/features/accessControl/components/NewCertificate.jsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+const mapDispatchToProps = (dispatch) => ({
+  submitForm: (data) => dispatch(actions.submitCertificateForm(data))
+})
+
+export default BaseNew.connect(
+  BaseNew.mapStateToProps('accessControl'),
+  mapDispatchToProps,
+  reduxForm({
```
Comment (jeffomatic) [inline dashboard/src/features/accessControl/components/NewCertificate.jsx]: Chris and I were experimenting with the following pattern, which keeps all the form stuff unified:

```
const form = reduxForm({
  form: 'newAccessGrantForm',
  fields: [
    'guardType',
    'subject[].key',
    'subject[].value',
    'policy',
  ],
  validate: values => {
    const errors = {}

    if (!values.policy) {
      errors.policy = 'Policy is required'
    }

    return errors
  },
})(NewCertificate)

export default BaseNew.connect(
  BaseNew.mapStateToProps('accessControl'),
  mapDispatchToProps,
  form
)
```

File: dashboard/src/features/accessControl/actions.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import TokenCreateModal from './components/TokenCreateModal'
+
+const baseActions = baseListActions('accessControl', {
+  clientApi: () => chainClient().accessControl
+})
+
+export default {
+  fetchItems: () => {
+    return (dispatch) => {
+      const promise = chainClient().accessControl.list()
```
Comment (dominic) [inline dashboard/src/features/accessControl/actions.js]: Good catch. This is repurposed from the old `baseListActions` version, which was more complex.

File: dashboard/src/features/accessControl/actions.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    const body = {...data}
+
+    body.guardType = 'access_token'
+
+    return function(dispatch) {
+      return chainClient().accessTokens.create({
+        id: body.guardData.id,
+        type: 'client', // TODO: remove me when deprecated!
+      }).then(tokenResp => {
+        chainClient().accessControl.create(body).then(grantResp => {
```
Comment (dominic) [inline dashboard/src/features/accessControl/actions.js]: No, but I prefer always returning :+1:

File: dashboard/src/features/accessControl/actions.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            state: {preserveFlash: true},
+          }))
+        })
+      }).catch(err => Promise.reject({_error: err}))
+    }
+  },
+
+  submitCertificateForm: data => {
+    const subject = data.subject
+    delete data.subject
```
Comment (dominic) [inline dashboard/src/features/accessControl/actions.js]: Refactor for multiple policies.

File: dashboard/src/features/accessControl/components/NewCertificate.jsx (hunk 2 above)
Comment (dominic) [inline dashboard/src/features/accessControl/components/NewCertificate.jsx]: That was my understanding, yeah. Internal isn't a type of token users can create.

**Comment by jeffomatic** [comment]
I'm getting the following error:

screen shot 2017-04-18 at 12 44 19 pm

**Pull request merged**